### PR TITLE
fix #6064 chore(nimbus): Cleanup of nimbus end-to-end tests

### LINKED ISSUE #6064: Make integration test setup cleaner using pytest fixtures
Right now we repeat a lot of steps within the nimbus e2e suite. We can reduce these by using fixtures, and a similar setup to the legacy e2e tests.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1548)

### PR DESCRIPTION
**Pull request by lennartschoch**
Because

* code in the nimbus e2e-tests is duplicated frequently (especially the initial experiment creation)

This commit

* moves the logic for the creation of experiments into a pytest fixture that can be re-used

### CODE REVIEW

**Comment by lennartschoch** [comment]
Oops, two of the e2e tests are also failing in CircleCI... so seems like I actually broke something.
Since there are more tests failing on my machine (3/4) than on the pipeline (2/4), I guess there is still something wrong with my setup though.

**Comment by lennartschoch** [comment]
Made some progress regarding this today:
- The tests on `main` are now passing on my machine. Seems like I just used the wrong .env file
- The tests on my branch are still failing. Weirdly, they still seem to fail because of `#approve-request-button` not being visible (which isn't a problem in CircleCI). I also discovered `make integration_vnc_up`, but since the test abort the moment they fail, it was hard to see what exactly is happening.

**Comment by jrbenny35** [comment]
Hey @lennartschoch! Thanks for working on this. For your local testing there is a few steps to run before the ```make integration_test_nimbus``` command will work. You can follow these commands https://github.com/mozilla/experimenter/blob/main/.circleci/config.yml#L77


Let me know if that helps.

**Comment by lennartschoch** [comment]
Hey @jrbenny35, thanks for the advice! I went through these steps yesterday and now the tests pass on `main` on my machine. For some reason, the test on my feature branch still fail for different reasons on my machine than on the pipeline. For example, the test seem to be unable to find the `#approve-request-button` in the nimbus UI on my machine, but on the pipeline the tests went past that point and only failed when approving or rejecting the experiment on Kinto.

**Comment by jrbenny35** [comment]
> 
> 
> Hey @jrbenny35, thanks for the advice! I went through these steps yesterday and now the tests pass on `main` on my machine. For some reason, the test on my feature branch still fail for different reasons on my machine than on the pipeline. For example, the test seem to be unable to find the `#approve-request-button` in the nimbus UI on my machine, but on the pipeline the tests went past that point and only failed when approving or rejecting the experiment on Kinto.

Gotcha, let me look through the PR and see if you missed something. The Approve button won't show if the experiment forms aren't filled out correctly.

**Comment by lennartschoch** [comment]
> > Hey @jrbenny35, thanks for the advice! I went through these steps yesterday and now the tests pass on `main` on my machine. For some reason, the test on my feature branch still fail for different reasons on my machine than on the pipeline. For example, the test seem to be unable to find the `#approve-request-button` in the nimbus UI on my machine, but on the pipeline the tests went past that point and only failed when approving or rejecting the experiment on Kinto.
> 
> Gotcha, let me look through the PR and see if you missed something. The Approve button won't show if the experiment forms aren't filled out correctly.

Oh, that's good to know! Maybe I was paying too much attention to the screen with the approve button and not enough attention to the experiment forms when trying to debug this. Will also have another look! I guess you're referring to the Approve button in Nimbus, not Kinto, correct?

**Comment by jrbenny35** [comment]
> 
> 
> > Gotcha, let me look through the PR and see if you missed something. The Approve button won't show if the experiment forms aren't filled out correctly.
> 
> Oh, that's good to know! Maybe I was paying too much attention to the screen with the approve button and not enough attention to the experiment forms when trying to debug this. Will also have another look! I guess you're referring to the Approve button in Nimbus, not Kinto, correct?

Hmm. Looks like the Experiment Not Found error is after it has been rejected on Kinto and should be showing on the drafts page. 

The other error points to a page not loading. Seems like the Summary Page isn't loading

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jrbenny35) [inline app/tests/integration/nimbus/test_e2e_create_experiment.py:39]: This could be a fixture and moved to ```conftest.py```.

**Review by jrbenny35 — CHANGES_REQUESTED** [review]
Some initial thoughts, thanks for starting on this!

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: The problem is that the experiment name of each test has to be different to make sure there are no conflicts between tests. I saw that there is a way of [parameterizing fixtures using pytest annotations](https://docs.pytest.org/en/latest/example/parametrize.html), but it looks a bit hacky to me. Should I still use it to define the experiment name for each test or do you have another idea?

**Comment by lennartschoch** [comment]
@jrbenny35 thanks for the feedback! :) I implemented all the changes requested by you except the one I [commented](https://github.com/mozilla/experimenter/pull/6099#discussion_r682172251) on.

Also, I managed to get the tests to pass both locally and on the pipeline. Still, I believe there might be something a bit flaky in the experiment creation part (at some point, different tests were failing locally and on CircleCI for different reasons, even though I was running the exact same commands on my machine that run on the pipeline).

 But for now everything seems to be fine! 🎉

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jrbenny35) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: Maybe concatenate some random numbers or a shortened UUID on the end of the assignment. 

```suggestion
data.public_name + UUID generation or something
```

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jrbenny35) [inline app/tests/integration/nimbus/test_e2e_create_experiment.py:51]: Love this

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jrbenny35) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: Do you still plan to move this to ```conftest.py```?

**Review by jrbenny35 — CHANGES_REQUESTED** [review]
This is looking good!

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: I do! Will take some time tomorrow to move this into a fixture and implement your random number / UUID idea!

Comment (jaredlockhart) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: I liked having the experiment name be the name of the test so it's easier to debug the tests after the fact?

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jaredlockhart) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:51]: What's being passed in here?  An instance of a data class?  Does mutating it share data between tests?

**Review by jaredlockhart — COMMENTED** [review]
Generally looking really good to me!  Some ideas for naming/structure.  Thanks for working on this @lennartschoch !

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jaredlockhart) [inline app/tests/integration/nimbus/test_e2e_create_experiment.py:111]: Oh I think my comment vanished somehow.  Are we mutating the same object that's passed to multiple tests?  Does this share data between tests?

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: @jaredlockhart I agree. In case we actually go for the generated experiment names, I think we would at least have to log them.

@jrbenny35 I just started putting the `create_experiment` function into a fixture. I think this might not be as easy as it seems. In some tests, we have steps that have to be executed **before** the creation of the experiment, e.g. counting the number of experiments in the `test_create_new_experiment_remote_settings` test or navigating to the Drafts tab in `test_create_new_experiment_remote_settings_reject`. Probably, in order to keep this behavior, we would need to move those parts of the code into fixtures as well. If the only reason for putting this function into a fixture is to make it available in multiple test files, I would suggest we don't create a fixture that executes `create_experiment`, but instead returns the `create_experiment` function itself, so we can execute it whenever we want. This pattern is also explained in the [pytest docs](https://docs.pytest.org/en/latest/how-to/fixtures.html#factories-as-fixtures). What do you think?

**Comment by lennartschoch** [comment]
Thanks for the feedback @jaredlockhart! Will create some commits based on it tomorrow :)

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:111]: This should not be the case. According to the [docs](https://docs.pytest.org/en/6.2.x/fixture.html#fixture-scopes), fixtures are destroyed after each test by default.

File: app/tests/integration/nimbus/models/base_dataclass.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/models/base_dataclass.py]: Fixed in https://github.com/mozilla/experimenter/pull/6099/commits/b71f79a2ed8b031ae43b0b75ca14f4359eab64b8

File: app/tests/integration/nimbus/models/base_dataclass.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/models/base_dataclass.py]: Fixed in https://github.com/mozilla/experimenter/pull/6099/commits/b71f79a2ed8b031ae43b0b75ca14f4359eab64b8

File: app/tests/integration/nimbus/conftest.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/conftest.py]: Fixed in https://github.com/mozilla/experimenter/pull/6099/commits/b71f79a2ed8b031ae43b0b75ca14f4359eab64b8

File: app/tests/integration/nimbus/conftest.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/conftest.py]: Fixed in https://github.com/mozilla/experimenter/pull/6099/commits/b71f79a2ed8b031ae43b0b75ca14f4359eab64b8

File: app/tests/integration/nimbus/conftest.py
Comment (lennartschoch) [reply app/tests/integration/nimbus/conftest.py]: One more thing I'd like to do here is using the value of the `channel` select instead of the name for consistency reasons. I'll wait for this until #6132 is resolved though.

**Comment by lmorchard** [comment]
A fix for #6132 landed, so I'll give CI a kick

**Comment by jaredlockhart** [comment]
@lennartschoch You should be able to rebase on `main` now to clear up the integration test failures 👍

File: app/tests/integration/nimbus/test_e2e_create_experiment.py
Comment (jrbenny35) [reply app/tests/integration/nimbus/test_e2e_create_experiment.py:11]: @lennartschoch I like that. Let's do that.

**Comment by jrbenny35** [comment]
@lennartschoch Hey! Do you have any time to work on this for the remainder of the week? This PR is blocking some tests I am looking to add, but I want to give you some time to finish it up. If you don't have the time I can add the final changes needed. Please let me know :)

**Comment by jaredlockhart** [comment]
This is looking really good @lennartschoch !  I'll let @jrbenny35 give the final sign off but it all looks right to me 👍 Thank you!

File: app/tests/integration/nimbus/conftest.py
Comment (jrbenny35) [inline app/tests/integration/nimbus/conftest.py:124]: Oh you went for extra points! Love it!

File: app/tests/integration/nimbus/conftest.py
Comment (jrbenny35) [inline app/tests/integration/nimbus/conftest.py:83]: Beautiful!

**Review by jrbenny35 — APPROVED** [review]
This is really great work here! R+, I hope you stick around for more! Awesome job!!

**Comment by jrbenny35** [comment]
@lennartschoch I'll give you the honors of hitting the ```squash and merge``` button.

**Comment by lennartschoch** [comment]
@jrbenny35 @jaredlockhart Thank you! :) It's a really cool product you created there, will definitely stick around! Unfortunately I don't have the permissions to merge though, so I'll leave that to one of you ;)

**Pull request merged**
